harness/determined/profiler.py

```diff
@@ -0,0 +1,517 @@
+import logging
+import queue
+import threading
+import time
+from typing import Optional, List, Dict, Any
+
+import psutil
+import pynvml
+
+
+from determined.common import api
+from determined.common.api import TrialProfilerMetricsBatch
+
+SYSTEM_METRIC_TYPE_ENUM = "PROFILER_METRIC_TYPE_SYSTEM"
+
+LOG_NAMESPACE = "determined-profiler"
+
+
+
+
+class ProfilerAgent:
+    """
+    Agent that collects metrics and sends them to the master. It has:
+    - a thread to send data to the Master API (sender_thread)
+    - a thread to collect System Metrics and periodically flush to sender_thread (sys_metric_collector_thread)
+    - [UNIMPLEMENTED] something to batch Timings and periodically flush to sender_thread (timings_batcher)
+
+    The ProfilerAgent needs to be created at the beginning of training and it needs
+    to be notified every time the batch_idx increases. When it is created, it launches
+    the sender_thread and the sys_metric_collector_thread. They will be cleaned up
+    either once end_after_batch is finished or after 5 minutes have passed.
+
+    You can also ship Timings through the ProfilerAgent with the record_timing() method. This
+    functionality has not yet been implemented - we need to batch the timings and periodically
+    flush them to the sender_thread.
+
+    Profiling is only active between start_on_batch and end_after_batch. It will also automatically
+    shut down 5 minutes after starting. When profiling is not active, no system metrics are collected
+    and the record_timing function is a no-op.
+
+    Usage:
+    ```
+    profiler_agent = ProfilerAgent(self, trial_id, agent_id, master_url, start_on_batch, end_after_batch)
+
+    for batch_idx, batch in enumerate(batches):
+        profiler_agent.update_batch_idx(batch_idx)
+
+        # NOTE: Timing API has not been fully developed yet
+        forward_pass_timing = Timing("forward_pass")
+        forward_pass_timing.start()
+        # Do forward pass
+        forward_pass_timing.end()
+        profiler_agent.record_timing(forward_pass_timing)
+    ```
+    """
+    def __init__(self, trial_id: int, agent_id: str, master_url: str, start_on_batch: int, end_after_batch: Optional[int] = None):
+        self.current_batch_idx = 0
+        self.agent_id = agent_id
+        self.trial_id = trial_id
+        self.master_url = master_url
+        self.start_on_batch = start_on_batch
+        self.end_after_batch = end_after_batch
+
+        # Track duration to stop collecting after 5 minutes. start_time also serves as
+        # the indicator that collection has begun.
+        self.start_time = None
+        # TODO: Currently we only check for shutdown in update_batch_idx(). This is a problem when there
+        #       is a long period of time between batch_idx being updated (such as during validation)
+        #       because it won't shut down until validation completes. We probably need to create a
+        #       new thread to enforce the timeout correctly.
```

Comment:
Why not just explicity shut down the collection when we transition from training -> validation?  Or hardcode the timeout into the collection thread rather than ProfilerAgent?